### ISSUE TITLE
Videos won't play (403 Forbidden) due to how nginx is configured

### ISSUE THREAD

**Issue by p0358** [tubearchivist/tubearchivist#268]
This is potentially the same issue as in #5, except I found the exact culprit.

The issue is that everything in the container runs as `root` (fine on its own), but nginx runs as `www-data`. So I ended up with an counter-intuitive situation, where the app would successfully download videos and its metadata, but attempting to play them back would result in a broken player in every browser. Opening webtools I've quickly learnt that the culprit is not the lack of codec, as Firefox reported in the UI, but error 403 from nginx. I've installed procps on the container and seen that nginx runs as www-data:

```
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root           1  0.0  0.0   5728  1704 ?        Ss   12:23   0:00 /bin/bash ./run.sh
root          16  0.2  0.4 122480 65772 ?        S    12:23   0:03 /usr/local/bin/python /root/.local/bin/celery -A home.tasks worker --loglevel=INFO
root          17  0.0  0.3 134616 65000 ?        S    12:23   0:00 /usr/local/bin/python /root/.local/bin/celery -A home beat --loglevel=INFO -s /cache/celerybeat-schedule
root          18  0.0  0.3 112976 54556 ?        S    12:23   0:00 uwsgi --ini uwsgi.ini
root          19  0.0  0.0  58352  1568 ?        Ss   12:23   0:00 nginx: master process nginx
www-data      20  0.0  0.0  59036  3788 ?        S    12:23   0:00 nginx: worker process
www-data      21  0.0  0.0  58684  2820 ?        S    12:23   0:00 nginx: worker process
www-data      22  0.0  0.0  58684  2820 ?        S    12:23   0:00 nginx: worker process
www-data      23  0.0  0.0  58684  2820 ?        S    12:23   0:00 nginx: worker process
root          27  0.1  0.3 121928 62712 ?        S    12:23   0:02 uwsgi --ini uwsgi.ini
root          29  0.0  0.3 122196 61216 ?        S    12:23   0:00 /usr/local/bin/python /root/.local/bin/celery -A home.tasks worker --loglevel=INFO
root          30  0.0  0.3 122200 61260 ?        S    12:23   0:00 /usr/local/bin/python /root/.local/bin/celery -A home.tasks worker --loglevel=INFO
root          31  0.0  0.3 122204 61264 ?        S    12:23   0:00 /usr/local/bin/python /root/.local/bin/celery -A home.tasks worker --loglevel=INFO
root          32  0.4  0.6 156724 98892 ?        S    12:23   0:07 /usr/local/bin/python /root/.local/bin/celery -A home.tasks worker --loglevel=INFO
root          69  0.0  0.0   2420   432 pts/0    Ss   12:42   0:00 /bin/sh
```

This is a problem since you cannot access `/youtube` as www-data, at least in my case:

```
www-data@tubearchivist-custom-app-5459b44875-mxgvq:/app$ ls /youtube/
ls: cannot open directory '/youtube/': Permission denied
```

and is counter-intuitive, since the rest of the app has no problem accessing the videos, you only cannot play them in the browser. I didn't try setting mount perms to 777 (which is a bad practice anyways), but since the files seem to be created with 000 permissions and owned as root, there's a chance this wouldn't work anyways (as was the case for the person in issue #5).

As such, instead of fiddling with external permissions and weakening the overall security, I think the best solution is to modify `/etc/nginx/nginx.conf` and changing `user www-data;` to `user root;` to make it consistent with the rest of the processes in the container. I did that and now playback works fine for me.

**Comment by p0358** [comment]
I am also getting this error:
```
2022-07-14T11:10:15.441022714Z [2022-07-14 13:10:15,440: WARNING/ForkPoolWorker-4] rSVepBAwwa8: get metadata from youtube
2022-07-14T11:10:16.673063093Z [2022-07-14 13:10:16,672: WARNING/ForkPoolWorker-4] UC-xALUOP3IO317Ufpl11GGw: get metadata from es
2022-07-14T11:10:16.745802345Z [2022-07-14 13:10:16,745: WARNING/ForkPoolWorker-4] rSVepBAwwa8: get sponsorblock timestamps
2022-07-14T11:10:16.968482823Z [2022-07-14 13:10:16,967: WARNING/ForkPoolWorker-4] rSVepBAwwa8: sponsorblock failed: Not Found
2022-07-14T11:10:16.968592263Z [2022-07-14 13:10:16,968: WARNING/ForkPoolWorker-4] rSVepBAwwa8-en: get user uploaded subtitles
2022-07-14T11:10:16.968650440Z [2022-07-14 13:10:16,968: WARNING/ForkPoolWorker-4] rSVepBAwwa8-pl: get user uploaded subtitles
2022-07-14T11:10:17.066062716Z [2022-07-14 13:10:17,065: ERROR/ForkPoolWorker-4] Task download_pending[0db292f6-0423-49d1-a7ae-b688fa238c61] raised unexpected: PermissionError(1, 'Operation not permitted')
2022-07-14T11:10:17.066114049Z Traceback (most recent call last):
2022-07-14T11:10:17.066134442Z   File "/usr/local/lib/python3.10/shutil.py", line 815, in move
2022-07-14T11:10:17.066175578Z     os.rename(src, real_dst)
2022-07-14T11:10:17.066193457Z OSError: [Errno 18] Invalid cross-device link: '/cache/download/rSVepBAwwa8_HORSE_TOUCHING_ASMR_tickles_your_ears.mp4' -> '/youtube/eat with anus man/20201117_rSVepBAwwa8_HORSE TOUCHING ASMR (tickles your ears ).mp4'
2022-07-14T11:10:17.066209660Z 
2022-07-14T11:10:17.066225654Z During handling of the above exception, another exception occurred:
2022-07-14T11:10:17.066257990Z 
2022-07-14T11:10:17.066274053Z Traceback (most recent call last):
2022-07-14T11:10:17.066289488Z   File "/root/.local/lib/python3.10/site-packages/celery/app/trace.py", line 451, in trace_task
2022-07-14T11:10:17.066305900Z     R = retval = fun(*args, **kwargs)
2022-07-14T11:10:17.066334954Z   File "/root/.local/lib/python3.10/site-packages/celery/app/trace.py", line 734, in __protected_call__
2022-07-14T11:10:17.066354579Z     return self.run(*args, **kwargs)
2022-07-14T11:10:17.066371201Z   File "/app/home/tasks.py", line 89, in download_pending
2022-07-14T11:10:17.066402629Z     downloader.run_queue()
2022-07-14T11:10:17.066436571Z   File "/app/home/src/download/yt_dlp_handler.py", line 190, in run_queue
2022-07-14T11:10:17.066452984Z     self.move_to_archive(vid_dict)
2022-07-14T11:10:17.066468628Z   File "/app/home/src/download/yt_dlp_handler.py", line 401, in move_to_archive
2022-07-14T11:10:17.066484482Z     shutil.move(old_file_path, new_file_path)
2022-07-14T11:10:17.066513046Z   File "/usr/local/lib/python3.10/shutil.py", line 835, in move
2022-07-14T11:10:17.066528830Z     copy_function(src, real_dst)
2022-07-14T11:10:17.066544614Z   File "/usr/local/lib/python3.10/shutil.py", line 435, in copy2
2022-07-14T11:10:17.066560468Z     copystat(src, dst, follow_symlinks=follow_symlinks)
2022-07-14T11:10:17.066592804Z   File "/usr/local/lib/python3.10/shutil.py", line 380, in copystat
2022-07-14T11:10:17.066608937Z     lookup("chmod")(dst, mode, follow_symlinks=follow)
2022-07-14T11:10:17.066621438Z PermissionError: [Errno 1] Operation not permitted: '/youtube/eat with anus man/20201117_rSVepBAwwa8_HORSE TOUCHING ASMR (tickles your ears ).mp4'
```

Weirdly enough the video does end up at its destination at /youtube eventually. It seems that the error `[Errno 18] Invalid cross-device link` is caused by /cache and /youtube being mounted at different underlying filesystems. I am not sure how the video still ends up at /youtube, but I know that this error causes the download queue to stop further downloads and the downloaded video not being removed from it as if it wasn't ever downloaded.

**Comment by bbilly1** [comment]
Thanks for looking into this. What OS/hardware are you running this on?

**Comment by p0358** [comment]
I am running TrueNAS Scale (so Linux + k3s), using setup instructions from your wiki that link there: https://heavysetup.info/applications/tube-archivist/about/ (custom app that allows to specify any docker containers to run)

It does seem like the person writing that tutorial also uses PVC for /cache on a different pool than mount for /youtube, so I'm not sure why they didn't run into the latter error (or perhaps they did but didn't notice?). As for the permission issue, they also mentioned they could only get mount to work with 777 permissions, but without elaborating much on what was the issue beyond that. For me it works just fine when I grant the `apps` user permissions to the path on host and also run container as root user, and TubeArchivist app can access it just fine, but the nginx running as a user that's neither root or what would match the `apps` user UID means it's unable to access the mount at all (also because "other" has its permissions set to 0 in my case).

As for the Python error, it seems to be this: https://stackoverflow.com/questions/42392600/oserror-errno-18-invalid-cross-device-link

Supposedly `shutil.move` in Python 3 should work fine with different filesystems, but people in the comments say they still get an error

**Comment by bbilly1** [comment]
Ah, TrueNAS. So I haven't done anything with that. @Heavybullets8 was kind enough to put the installation instructions together for that. 

This is not an issue that happens with regular Docker but I remember in the very beginning before containerizing the application, I used `os.remane` which broke in the container as these folders are now volumes and mount points there. But this already uses `shutil.move` now: https://github.com/tubearchivist/tubearchivist/blob/1479516689badaea804ab9bda52b4f27bdf5109b/tubearchivist/home/src/download/yt_dlp_handler.py#L401.

So not sure what can be done here. But I assume this is related to the nginx permission, something on TrueNas is setting additional file attributes to folders? Or maybe a parent folder that blocks access?

In my container */youtube* has `755` permissions. That should be enough as nginx only needs to read and serve these files, the application creates, modifies or deletes them.

I'm not sure if running nginx - that serves the whole application and is potentially exposed to the internet - as root in the container is the smart play here, or at least, if it is the better option than to set 777 on the media volume. Especially as this seams to be a TrueNAS specific issue.

**Comment by p0358** [comment]
The Python error is potentially a separate issue from the nginx issue, but I'm also very confused on why it happens.

> In my container /youtube has 755 permissions. That should be enough as nginx only needs to read and serve these files, the application creates, modifies or deletes them.

It seems that the mount points in TrueNAS directly inherit all permissions and owner uid/gid from the real filesystem. I cannot run `chmod 755 /youtube` from within the container.

Btw doesn't your host path also have 755 permissions? I think the fact it works in the default setup might be a mere coincidence, where folders by default in Linux distros get created with 755 permissions for dirs and 644 for files, making the default setup coincidentally just work for most people: https://ikriv.com/blog/?p=4698

> I'm not sure if running nginx - that serves the whole application and is potentially exposed to the internet - as root in the container is the smart play here, or at least, if it is the better option than to set 777 on the media volume

I think it is honestly. My reasoning is that it runs in the container anyways, and is meant to have full access to the downloaded videos anyways, so if it's somehow exploited and compromised, the scope of access should theoretically be the same. The only thing though that could be leaked is the YouTube cookie, that's assuming that in current setup nginx wouldn't have access to it (but at a glance it seems it does). Also nginx reverse-proxies requests to the app running on root anyways, meaning that any potential vulnerability that doesn't involve broken HTTP requests would still be relied to the app anyways.

Well, I'm not an expert on Docker to be honest, but I'd assume the configurability of the user the container runs at is also for the purpose of ensuring FS permissions work fine, with the expectation that "one container = one app", that is also running as the same user (unless some subservices don't access FS and then can drop permissions?)

The only other solution I could theoretically use, that doesn't involve container changes, would be attempting to change the host directory owner from `apps` user to the numerical ID of www-data of the container, but it feels like an ugly workaround that would also break other containers running as `apps` user accessing the same dataset/folder.

**Comment by p0358** [comment]
> I cannot run `chmod 755 /youtube` from within the container.
```
[...]
During handling of the above exception, another exception occurred:
[...]
2022-07-14T11:10:17.066592804Z   File "/usr/local/lib/python3.10/shutil.py", line 380, in copystat
2022-07-14T11:10:17.066608937Z     lookup("chmod")(dst, mode, follow_symlinks=follow)
```

This might potentially explain the Python error though.

I looked into how shutil.py is implemented, and it tries to move, catches the error, and then attempts copy. Unfortunately during copy it attempts to `chmod` (because it also attempts to copy all possible filesystem metadata), and the `chmod` call isn't under try-catch block expecting to fail for reason other than not being implemented. This explains why the file gets copied, but an error is thrown afterwards. This also explains `During handling of the above exception, another exception occurred:`, because the fallback copy happens from within the catch block of the first exception

**Comment by p0358** [comment]
Potential solution for Python error:
```py
def move(src, dst, copy_function=copy2):
    """
[...]
    The optional `copy_function` argument is a callable that will be used
    to copy the source or it will be delegated to `copytree`.
    By default, copy2() is used, but any function that supports the same
    signature (like copy()) can be used.
```

```py
def copy(src, dst, *, follow_symlinks=True):
    """Copy data and mode bits ("cp src dst"). Return the file's destination.
    The destination may be a directory.
    If follow_symlinks is false, symlinks won't be followed. This
    resembles GNU's "cp -P src dst".
    If source and destination are the same file, a SameFileError will be
    raised.
```

```py
def copy2(src, dst, *, follow_symlinks=True):
    """Copy data and metadata. Return the file's destination.
    Metadata is copied with copystat(). Please see the copystat function
    for more information.
```

I think the solution is to specify the optional argument `copy_function` in the call to `move` to use the function `copy` instead of `copy2`

**Comment by p0358** [comment]
I've managed to reproduce the nginx issue by mounting the docker-compose.yml's volume into a host dir on a Debian host and chmodding it to 700, also managing to fix it by running nginx as root, confirming my earlier suspicions (despite the permissions on files being set to 755 which should in theory make them readable by anyone...), I think I'll submit a PR to fix this.

**Comment by Heavybullets8** [comment]
@p0358 would you mind messaging me on discord (heavybullets#5535) or opening an issue on https://github.com/Heavybullets8/HeavySetup

For me, not setting the perms for the dataset as 777 resulted in the video showing up in the catalog, but if I clicked on the video, then clicked play, nothing would happen. (Maybe there was a playback error, I can't remember). But changing the perms to 777 for the dataset fixed this, and I've had no issues since. (This is also the case for a few other apps, Truenas SCALE has been a pain when it comes to application permissions)

If there is a better method to setting up an application, I'd like to update the documentation to reflect that, as well as make those changes for my own application lol. Thank you.

**Comment by bbilly1** [comment]
Thank you very much for your thorough research on this topic. Some comments:

> The only thing though that could be leaked is the YouTube cookie  

I don't think that's an issue. The cookie never gets read by nginx. It gets stored in Redis and will get read in memory by the application only for the yt-dlp calls, it never gets return by any view. So as far as I can tell, there shouldn't be any way how this could leak. Or what makes you think that could be a problem?

> one container = one app  

I know, that's a common container philosophy. But we are quite far a way from that. There are even regularly people complaining that they need to install three containers. :-)

So all in all, I think your solution in #270 makes sense.

**Comment by p0358** [comment]
Fully fixed with 146f17b
